hoomd/mpcd/SphereGeometry.h

```diff
@@ -0,0 +1,195 @@
+         *       amount of time remaining.
+         */
+        HOSTDEVICE bool detectCollision(Scalar3& pos, Scalar3& vel, Scalar& dt) const
+            {
+
+            /*
+             * Detect if particle has left the box, and try to avoid branching or absolute value calls. The sign used
+             * in calculations is 1 if the particle is out-of-bounds in the radial direction, 0 otherwise.
+             *
+             * We intentionally use > / < rather than >= / <= to make sure that spurious collisions do not get detected
+             * when a particle is reset to the boundary location. A particle landing exactly on the boundary from the bulk
+             * can be immediately reflected on the next streaming step, and so the motion is essentially equivalent up to
+             * an epsilon of difference in the channel width.
+             */
+            double r2 = pos.x*pos.x + pos.y*pos.y + pos.z*pos.z;
+            const char sign = (r2 > R*R);
+            // exit immediately if no collision is found or particle is not moving normal (radial direction) to the wall
+            // (since no new collision could have occurred if there is no normal motion)
+            /*
+            r(t) = sqrt(x**2 + y**2 + z**2)
+            radial velocity, r_dot = (x*x_dot + y*y_dot + z*z_dot)/r(t)
+            */
+            float vr = (pos.x*vel.x + pos.y*vel.y + pos.z*vel.z)/sqrt(r2);
+            if (sign == 0 || vr == Scalar(0))
+               {
+               dt = Scalar(0);
+               return false;
+               }
+
+            // Find the point of contact when the particle is just leaving the surface
+            /*
+            * Calculating the time spent outside-bounds requires the knowledge of the point of contact on the boundary.
+            * The point of contact can be calculated using geometrical considerations.
+            *
+            * We know the following quantities,
+            *    1. r(t+del_t)      (the point outside the sphere)
+            *    2. v(t)            (previous velocity)
+            *
+            * Assuming 'r*' is the point of contact on the spherical shell and dx* is the distance travelled by the
+            * particle outside the boundary.
+            *
+            * r* = r(t+del_t) + dx* * v(t)/|v|          (the -ve signs cancel out in the 2nd term)
+            * and |r|**2 = R**2
+            *
+            * Solving the above equations,
+            * dx* = -r(t+del_t).v(t)/|v| +- sqrt((r(t+del_t).v(t))**2 - |r(t+del_t)|**2 + R**2)
+            * consequently, dt = dx* / |v|
+            */
+            double mod_v,r_dot_v_hat,dx_;
+
+            mod_v = sqrt(vel.x*vel.x + vel.y*vel.y + vel.z*vel.z);
+            r_dot_v_hat = (pos.x*vel.x + pos.y*vel.y + pos.z*vel.z)/mod_v;
+
+            dx_ = -r_dot_v_hat + sqrt(r_dot_v_hat*r_dot_v_hat - r2 + R*R);
+            dt = dx_/mod_v;
```

Comment:
Is there not a way to solve for dt directly, rather than dx? I didn't check your math, but it would be better to get dt since this is what is used below.

Reply:
I think this is the only way due to the geometry we are working with. I will also write it down properly and add a picture of that in this thread so that it is easier to understand and convenient to refer later on.

Reply:
https://seafile.rlp.net/f/3c48571cc6194385b6a8/?dl=1